### DISCUSSION TITLE
xmake import模块失败问题

### DISCUSSION THREAD

**Discussion by peter-233** [xmake-io/xmake#1916] [Q&A] [answered]
我试图用xmake导入libtorch，因此我仿照可以跑通的CMakeLists.txt写了一段程序：

```jsx
import("lib.detect.find_package")
local package = find_packages("cmake::Torch", {envs = {CMAKE_PREFIX_PATH = "/home/peter-233/Downloads/libtorch-cxx11-abi-shared-with-deps-1.10.0+cpu/libtorch"}})

target("TorchTest")
    add_cxxflags("${TORCH_CXX_FLAGS}") 
    add_packages("Torch")
    add_files("*.cpp")
    add_rules("mode.release", "mode.debug")
    set_kind("binary")
```

但是，当我运行xmake后，他显示了如下错误：

```jsx
error: ./xmake.lua:2: attempt to call global 'import' (a nil value)
stack traceback:
    [./xmake.lua:2]: in main chunk

checking for platform ... linux
checking for architecture ... x86_64
error: ./xmake.lua:2: attempt to call global 'import' (a nil value)
stack traceback:
    [./xmake.lua:2]: in main chunk
```

请问我该如何解决？

**Comment by waruqi** [comment, marked as answer]
关于这块，之前文档写的不明确，find_package 目前只能在脚本域使用，虽然也可以集成包，但是用法比较繁琐

现在不管是远程包，还是本地包，还是系统库，都是统一走 add_requires 集成，用法更加简单一致

关于 add_requires 对于第三方库的安装集成，可以看文档：https://xmake.io/#/zh-cn/package/remote_package?id=%e7%ac%ac%e4%b8%89%e6%96%b9%e4%be%9d%e8%b5%96%e5%8c%85%e5%ae%89%e8%a3%85

而 cmake 包，2.6.1 线上版本，add_requires 对它的处理还不是很好，不支持透传 configs ，但基础的集成也是可以用的。

我更新下了下文档，看下 https://xmake.io/#/zh-cn/package/local_package?id=%e5%9c%a8%e9%a1%b9%e7%9b%ae%e4%b8%ad%e9%9b%86%e6%88%90%e5%8c%85

如果你的 cmake::xxx 包用到了一些组件, envs 配置什么的，需要更新到 xmake 的 dev 版本，`xmake update -s dev` 才能支持上

用法见: https://xmake.io/#/zh-cn/package/local_package?id=%e6%8c%87%e5%ae%9a%e7%bb%84%e4%bb%b6

关于这块的相关 issues 见：https://github.com/xmake-io/xmake/issues/1632#issuecomment-995950024

**Reply by peter-233** [reply]
1. 我首先仿照文档尝试了add_requires的方法，xmake.lua如下：
```jsx
add_requires("cmake::Torch", {alias = "pytorch", system = true, configs = {envs = {CMAKE_PREFIX_PATH = "/home/peter-233/Downloads/pytorch/libtorch/"}}})
target("TorchTest")
    add_packages("pytorch")
    add_files("*.cpp")  
    add_rules("mode.release", "mode.debug")
    set_kind("binary")
```
在xmake后，他提示我如下信息：
```jsx
warning: target(TorchTest).add_linkdirs("-Wl,--no-as-needed,/home/peter-233/Downloads/pytorch/libtorch/lib") path not found at :-1
```
可是我十分确定该路径正确无误，pwd显示路径：/home/peter-233/Downloads/pytorch/libtorch/lib
运行xmake r后，他提示我找不到动态链接库，请问该如何解决？

2. 此外，我在使用cmake进行编译时使用的命令为：cmake -DCMAKE_PREFIX_PATH=/home/peter-233/Downloads/pytorch/libtorch .. ，我在xmake中是否能通过不在add_requires中加config而是直接add_defines(“CMAKE_PREFIX_PATH = ...”)来达到相同的效果？

**Reply by waruqi** [reply]
> warning: target(TorchTest).add_linkdirs("-Wl,--no-as-needed,/home/peter-233/Downloads/pytorch/libtorch/lib") path not found at :-1

这个是因为提取的 cmake 包信息中带了 -Wl,--no-as-needed, 错误识别成了 linkdirs ，我稍微改了下，更新下 dev 再试试

> 此外，我在使用cmake进行编译时使用的命令为：cmake -DCMAKE_PREFIX_PATH=/home/peter-233/Downloads/pytorch/libtorch .. ，我在xmake中是否能通过不在add_requires中加config而是直接add_defines(“CMAKE_PREFIX_PATH = ...”)来达到相同的效果？

不能，你没理解这里面的机制，add_requires("cmake::xxx") 内部会去自动生成一个 CMakeLists.txt 去拼接 find_package ，然后调用 cmake 命令去提取 cmake 里面的包信息，所以这个 CMAKE_PREFIX_PATH 是调用 cmake 命令时候，设置的环境变量，让 cmake 输出 libtorch 信息时候，能够正确读取到

而 add_defines 是给正常代码加 宏定义，完全两码事，不要混一起。

**Reply by peter-233** [reply]
非常感谢，目前我的代码可以跑通了。此外我还有几个关于xmake的问题：
1. 不太明白add_requires()和add_package()的区别在哪里？我的感觉是add_require()类似于把cmake中find_package封装了，add_package类似于把cmake中link_libraries和include_directories封装了，这样理解正确么？
2. add_requires(“xxx”)，这个xxx是从哪里找的？比如说find_package("xxx")，我知道是从Findxxx.cmake中找的，那add_requires(“xxx”)中的包又是从哪里找的？我在试这个demo的时候发现xxx是Torch或是torch不改变输出信息，xxx是不区分大小写的么？
3. 我是从libTorch的官网下载的包，系统是ubuntu18.04。现在每次都要add_requires(cmake::Torch ， {envs = ...})这样才能运行，我能否进行一些操作使得我之后写xmake.lua的时候只需要写add_requires(Torch)然后add_package()就能解决问题?比如设置一些环境变量什么的？

**Reply by waruqi** [reply]
> * 不太明白add_requires()和add_package()的区别在哪里？我的感觉是add_require()类似于把cmake中find_package封装了，add_package类似于把cmake中link_libraries和include_directories封装了，这样理解正确么？

add_requires 只是全局的依赖包声明，告诉项目 我需要哪些包，xmake 会有些走 find_package 从系统中找，找不到，再走安装逻辑，帮你安装

但是你的项目可能有很多 target，不是每个target 都需要这些包，所以需要 add_packages 对特定的 target 绑定需要的包，自动帮你对接 links linkdirs includedirs 等包配置。。

> add_requires(“xxx”)，这个xxx是从哪里找的？比如说find_package("xxx")，我知道是从Findxxx.cmake中找的，那add_requires(“xxx”)中的包又是从哪里找的？我在试这个demo的时候发现xxx是Torch或是torch不改变输出信息，xxx是不区分大小写的么？

不带 :: 域的，都是xmake-repo 的仓库的官方包，例如 add_requires("zlib")  https://github.com/xmake-io/xmake-repo/blob/master/packages/z/zlib/xmake.lua

带 cmake::zlib, vcpkg::zlib conan::zlib 这种的，都是第三方包管理的包，会自动调用对方的命令，走安装或者查找。文档都有详细说明：https://xmake.io/#/zh-cn/package/remote_package?id=%e7%ac%ac%e4%b8%89%e6%96%b9%e4%be%9d%e8%b5%96%e5%8c%85%e5%ae%89%e8%a3%85

官方包，都是小写的，第三方包，根据不同的包管理器的包命令规范而定，xmake不会去约定他们，因为包在它们那里。

> 我是从libTorch的官网下载的包，系统是ubuntu18.04。现在每次都要add_requires(cmake::Torch ， {envs = ...})这样才能运行，我能否进行一些操作使得我之后写xmake.lua的时候只需要写add_requires(Torch)然后add_package()就能解决问题?比如设置一些环境变量什么的？

自己本地定义一个 package("torch")，然后内部走 package:find_package("cmake::xxx", {configs = {envs = {}}}) 去包装下，参考：https://github.com/xmake-io/xmake-repo/blob/master/packages/g/gtk+3/xmake.lua

**Reply by peter-233** [reply]
我已经成功通过本地定义的package引入了，还有两个问题想请教一下：
1. 我在libtorch中加入了xmake.lua如下：
```jsx
package("pytorch")
    on_fetch("linux", function(package, opt)
        if opt.system then
            return find_package("cmake::Torch", {system = true, configs = {envs = {CMAKE_PREFIX_PATH = "/home/peter-233/project/torchtest/libtorch"}}})
        end
    end)
```
这里on_fetch的触发时机是什么？就是说我在文档中看到有很多对应package的on_XXX接口，他们是在我add_require()这些库的时候被执行的么？区分这些on_fetch()，on_build()是因为他们触发时机不同，要在这些时机对包做不同处理，这是他们的主要区别么？

2. 我在下载一个包的时候，比如说从官网下载了一个包，我可以通过哪些途径知道我该add_require()的包名字是什么？像这种情况只能去官网查阅是么？会不会有文件记录包的名字这种东西？

**Reply by waruqi** [reply]
用 package:find_package 不要用 find_package，在包里面

> 这里on_fetch的触发时机是什么？就是说我在文档中看到有很多对应package的on_XXX接口，他们是在我add_require()这些库的时候被执行的么？区分这些on_fetch()，on_build()是因为他们触发时机不同，要在这些时机对包做不同处理，这是他们的主要区别么？

安装依赖包前的 系统库查找，安装后的 查找

> 我在下载一个包的时候，比如说从官网下载了一个包，我可以通过哪些途径知道我该add_require()的包名字是什么？像这种情况只能去官网查阅是么？会不会有文件记录包的名字这种东西？


xrepo search xxx 或者直接到 xmake-repo 仓库里面翻下，或者到 xrepo.xmake.io 站点找下

**Reply by peter-233** [reply]
我修改的xmake.lua如下：
```jsx
package("pytorch")
    on_fetch("linux", function(package, opt)
        if opt.system then
            return package:find_package("cmake::Torch", {system = true, configs = {envs = {CMAKE_PREFIX_PATH = "/home/peter-233/project/torchtest/libtorch"}}})
        end
    end)
```
但是xmake后提示我：
```jsx
error: The packages(pytorch) not found, please run `xmake repo -u` first!
```
在xmake repo -u 后重新xmake依然报错，请问如何解决？

**Reply by waruqi** [reply]
你的包定义放在哪？

**Reply by peter-233** [reply]
我的demo结构如下：
```
├── 6.jpg
├── best.pkl
├── build
│   └── linux
├── CMakeLists.txt
├── jitmodel.pth
├── libtorch
│   ├── bin
│   ├── build-hash
│   ├── build-version
│   ├── include
│   ├── lib
│   ├── share
│   └── xmake.lua
├── model.pth
├── testsubdir
│   ├── subdir.cpp
│   └── xmake.lua
└── xmake.lua
```
其中libtorch就是我要用的包，项目根目录的xmake.lua include了libtorch目录和testsubdir目录，main函数在subdir.cpp

**Reply by waruqi** [reply]
我是要看你的 package("pytorch") 放在哪

**Reply by peter-233** [reply]
放在libtorch中的xmake.lua

**Reply by waruqi** [reply]
那不可能找不到 pytorch 包，除非你 includes 不对，我这测试可以

**Reply by peter-233** [reply]
我又重新试了一下，不加package:就可以正常运行，会不会是我其他xmake.lua写的有问题，麻烦您看一下，非常感谢。
主项目中xmake.lua:
```jsx
add_requires("opencv", "pytorch")
includes("libtorch", "testsubdir")
target("TorchTest")
    add_packages("pytorch", "opencv")
    -- add_files("*.cpp")  
    add_rules("mode.release", "mode.debug")
    set_kind("binary")
```

testsubdir中xmake.lua:
```jsx
target("TorchTest")
    add_files("subdir.cpp")
```

libtorch中xmake.lua:
```jsx
package("pytorch")
    on_fetch("linux", function(package, opt)
        if opt.system then
            return find_package("cmake::Torch", {system = true, configs = {envs = {CMAKE_PREFIX_PATH = "/home/peter-233/project/torchtest/libtorch"}}})
        end
    end)
```

**Reply by waruqi** [reply]
给个完整可以复现的 demo ，作为附件上传上来

**Reply by peter-233** [reply]
附件太大没法上传，没压缩小1G，压缩了200+MB，因为我是直接把libtorch包放在项目里面了。我发到您邮箱可以么？

**Reply by waruqi** [reply]
> 附件太大没法上传，因为我是直接把libtorch包放在项目里面了。我发到您邮箱可以么？

这个错误只是xmake.lua配置问题，二进制库可以不用传，就传配置

**Reply by peter-233** [reply]
[torchtest.zip](https://github.com/xmake-io/xmake/files/7736533/torchtest.zip)

**Reply by waruqi** [reply]
只是由于 package:find_package 里面你传递的 configs 参数，被内部覆盖了，我刚更新修复了。。可以更新到 dev 再试下，或者暂时先直接用 find_package 也可以

**Reply by peter-233** [reply]
我更新到dev并成功编译运行了我的程序。非常感谢！
